#Outline

In this colab, we study how to handle **large-scale datasets** in sklearn.  

* In this course, so far we were able to load **entire data in memory** and were able to train and make inferences on all the data at once.  

* The large scale data sets may not fit in memory and we need to devise strategies to handle it in the context of training and prediction use cases. 

In this colab, we will discuss the following topics:
> - Overview of handling large-scale data
- Incremental preprocessing and learning.
  -  `fit()` vs. `partial_fit()`: `partial_fit` is our friend in this case.
- Combining preprocessing and incremental learning.

# **Large-scale Machine Learning**

Large-scale Machine Learning differs from traditional machine learning in the sense that it involves processing large amount of data in terms of its **size** or **number of samples**, **features** or **classes**.

There were many exciting developments in efficient large scale learning on many real world use cases in the last decade.

Although scikit-learn is optimized for **smaller data**, it does offer a decent set of **feature preprocessing** and **learning algorithms** such as classification, regression and clustering for large scale data . 

Scikit-learn handles large data through `partial_fit()` method instead of using the usual `fit()` method. 
> The idea is to process data in **batches** and **update** the model parameters for each batch.  This way of learning is referred to as '**Incremental (or out-of-core) learning**'.

##Incremental Learning

Increamental learning may be required in the following two scenarios:

* For **out-of-memory (large) datasets**, where it’s not possible to **load the entire data into the RAM** at once, one can load the data in chunks and fit the training model for each chunk of data.

* For machine learning tasks where a new batch of data comes with time, re-training the model with the previous and new batch of data is a computationally expensive process. 
> Instead of re-training the model with the entire set of data, one can employ an incremental learning approach, where the model parameters are updated with the new batch of data.


###Incremental Learning in `sklearn`

To perform incremental learning, Scikit-learn implements **`partial_fit`** method that helps in training an out-of-memory dataset. In other words, it has the ability to learn incrementally from a batch of instances.

In this colab, we will see an example of how to read, process, and train on such a large dataset that can't be loaded in memory entirely. 

This method is expected to be called several times consecutively on different chunks of a dataset so as to implement out-of-core (online) learning. This function has some performance overhead, so it’s recommended to call it on a considerable large batch of data (that fits into the memory), to overcome the limitation of overhead.

### partial_fit() attributes:

`partial_fit(X, y, [classes], [sample_weight])`

where,

* `X` : array of shape (n_samples, n_features) where n_samples is the number of samples and n_features is the number of features.

* `y` : array of shape (n_samples,) of target values.

* `classes` : array of shape (n_classes,) containing a list of all the classes that can possibly appear in the y vector.

Must be provided at the first call to partial_fit, can be omitted in subsequent calls.

* `sample_weight` : (optional) array of shape (n_samples,) containing weights applied to individual samples (1. for unweighted).

Returns: object (self)

For classification tasks, we have to pass the list of possible target class labels in `classes` parameter to cope-up with the unseen target classes in the 1st batch of the data.

The following estimators implement `partial_fit` method:
* **Classification:** 
  * `MultinomialNB`
  * `BernoulliNB`
  * `SGDClassifier`
  * `Perceptron`

* **Regression:** 
  * `SGDRegressor`

* **Clustering:** 
  * `MiniBatchKMeans`


`SGDRegressor` and `SGDClassifier` are commonly used for handling large data. 

The problem with standard regression/classification implementations such as batch gradient descent, support vector machines (SVMs), random forests etc is that because of the need to load all the data into memory at once, they can not be used in scenarios where we do not have sufficient memory. SGD, however, can deal with large data sets effectively by breaking up the data into chunks and processing them sequentially. The fact that we only need to load one chunk into memory at a time makes it useful for large-scale data as well as cases where we get streams of data at intervals. 

#**fit() versus partial_fit()**

Below, we show the use of `partial_fit()` along with `SGDClassifier` on a sample data.

For illustration, we first use traditional `fit()` method and then use `partial_fit()' on the same data. 

In [1]:
# Importing Libraries
from sklearn.linear_model import SGDClassifier
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

##1. Traditional Approach (using `fit()`)

###Sample dataset

We will use a synthetic classification dataset for demonstration. 

Let us have 50000 samples with 10 features in the feature matrix. Further, lets have 3 classes in the target label, each class having a single cluster.

In [2]:
x, y = make_classification(n_samples=50000, n_features=10, 
                           n_classes=3, 
                           n_clusters_per_class=1)
xtrain, xtest, ytrain, ytest = train_test_split(x, y, test_size = 0.15)


We will make use of `SGDClassifier` to learn the classification model.

In [3]:
clf1 = SGDClassifier(max_iter=1000, tol=0.01)


We will use traditional `fit()` method to train our model.

In [4]:
clf1.fit(xtrain, ytrain)

SGDClassifier(tol=0.01)

Let's obtain the training and test scores on the trained model.

In [5]:
train_score = clf1.score(xtrain, ytrain)
print("Training score: ", train_score) 


Training score:  0.928


In [6]:
test_score = clf1.score(xtest, ytest)
print("Test score: ", test_score)

Test score:  0.9328


We obtain the confusion matrix and classification report for evaluating the classifier.

In [7]:
ypred = clf1.predict(xtest)

cm = confusion_matrix(ytest, ypred)
print(cm) 

[[2206   50  205]
 [  20 2311  136]
 [  27   66 2479]]


We use `classification_report` API for obtaining important evaluation metrics for all three classes.

In [8]:
cr = classification_report(ytest, ypred)
print(cr) 

              precision    recall  f1-score   support

           0       0.98      0.90      0.94      2461
           1       0.95      0.94      0.94      2467
           2       0.88      0.96      0.92      2572

    accuracy                           0.93      7500
   macro avg       0.94      0.93      0.93      7500
weighted avg       0.94      0.93      0.93      7500



#2. Incremental approach (using partial_fit())

We will now assume that the data can not be kept completely in the main memory and hence, will load chunks of data and fit using `partial_fit()`.

In [9]:
xtrain[0:5]

array([[ 2.13922458,  4.12443203,  0.18652789,  3.01372967, -1.9062114 ,
         0.81998862,  2.90943478, -2.38328903, -0.11806898, -0.82274741],
       [-1.07858661, -2.04420725,  0.80905554, -1.15324404,  1.52693968,
         1.66380706, -0.88203576, -0.05158305,  1.74644674, -0.92044429],
       [ 0.73819169,  1.40298284,  0.05266006,  0.82988662, -0.9823266 ,
         1.23638737, -2.31362451, -1.81441346,  0.5266005 ,  0.88605074],
       [ 0.39031529,  0.70154232, -0.28749983,  0.02101712, -1.16482668,
         0.67419337, -1.148984  ,  1.29178847,  0.10985247,  0.72437227],
       [ 0.35975069,  0.80329139, -0.45071496,  1.64457563,  1.43714942,
         0.63238187,  1.46047735,  0.4075769 ,  0.04767075, -0.61074532]])

In [10]:
ytrain[0:5]

array([0, 1, 0, 0, 1])



In order to load data chunk by chunk, we will first store the given (training) data in a csv file. (This is just for demonstration purpose. In a real scenario, the large dataset might already be in the form of say, a csv, which we will be reading in multiple iterations.)

In [11]:
import numpy as np

In [12]:
train_data = np.concatenate((xtrain, ytrain[:, np.newaxis]), axis=1)

In [13]:
train_data[0:5]

array([[ 2.13922458,  4.12443203,  0.18652789,  3.01372967, -1.9062114 ,
         0.81998862,  2.90943478, -2.38328903, -0.11806898, -0.82274741,
         0.        ],
       [-1.07858661, -2.04420725,  0.80905554, -1.15324404,  1.52693968,
         1.66380706, -0.88203576, -0.05158305,  1.74644674, -0.92044429,
         1.        ],
       [ 0.73819169,  1.40298284,  0.05266006,  0.82988662, -0.9823266 ,
         1.23638737, -2.31362451, -1.81441346,  0.5266005 ,  0.88605074,
         0.        ],
       [ 0.39031529,  0.70154232, -0.28749983,  0.02101712, -1.16482668,
         0.67419337, -1.148984  ,  1.29178847,  0.10985247,  0.72437227,
         0.        ],
       [ 0.35975069,  0.80329139, -0.45071496,  1.64457563,  1.43714942,
         0.63238187,  1.46047735,  0.4075769 ,  0.04767075, -0.61074532,
         1.        ]])

In [14]:
a = np.asarray(train_data)
np.savetxt("train_data.csv", a, delimiter=",")

Now, our data for demonstration is ready in a csv file. 

Let's create `SGDClassifier` object that we intend to train with `partial_fit`.

In [15]:
# Let us create another classifier and we will fit it incrementally.
clf2 = SGDClassifier(max_iter=1000, tol=0.01)


###Processing data chunk by chunk

Pandas' `read_csv()` function has an attributre `chunksize` that can be used to read data chunk by chunk. The `chunksize` parameter specifies the number of rows per chunk. (The last chunk may contain fewer than chunksize rows, of course.)

We can then use this data for `partial_fit`. We can then repeat these two steps multiple times. That way, entire data may not be reqiuired to be kept in memory.

In [16]:
import pandas as pd

chunksize = 1000

iter = 1
for train_df in pd.read_csv("train_data.csv", chunksize=chunksize,
                            iterator=True):
  if iter == 1:
    # In the first iteration, we are specifying all possible class 
    # labels.
    xtrain_partial = train_df.iloc[:, 0:10]
    ytrain_partial = train_df.iloc[:, 10]
    clf2.partial_fit(xtrain_partial, ytrain_partial,
                     classes=np.array([0, 1, 2]))
  else:
    xtrain_partial = train_df.iloc[:, 0:10]
    ytrain_partial = train_df.iloc[:, 10]
    clf2.partial_fit(xtrain_partial, ytrain_partial)

  print("After iter #", iter)
  print(clf2.coef_)
  print(clf2.intercept_)  
  iter = iter + 1

After iter # 1
[[   7.04991821   14.67800392   -5.6950819    21.19363375   11.11613728
     4.32880561   21.18274499    0.90294732   15.14595315    6.01100557]
 [  -8.88600215  -12.02918423    5.69217809   40.41252555   89.69079156
     7.28968813    1.30046203    7.09430256   -5.44765125   -6.57433245]
 [   0.75817665   -4.96298856    8.60580686  -65.57218862 -103.62751615
     4.46002317    1.94775891    1.2688729    14.43362046   -0.27348332]]
[ -28.39262248   -8.02563635 -115.90762174]
After iter # 2
[[  4.11632848   7.73006027   6.82912507   3.65991144  -6.97271659
   -7.5272327  -10.70776704   2.22706072   2.67977216 -16.36346034]
 [ -8.24308768 -11.84350147   0.08865239  30.3871914   72.23060942
    4.77743783   0.39751972  15.87808136  -1.20602893  11.34220791]
 [  0.78332802  -3.69668664 -12.87610311 -52.90511101 -84.13542984
  -11.14936079  -8.69066706  -6.11355121   7.53548307   5.02082049]]
[  -8.07151847   -6.93243287 -113.00258008]
After iter # 3
[[  5.30853154  10.348827

**Notes:** 

* In the first call to `partial_fit()`, we passed the list of possible target class labels. For subsequent calls to `partial_fit()`, this is not required.

* Observe the changing values pf the classifier attributes: `coef_` and `intercept_` which we are printing in each iteration.

In [17]:
test_score = clf2.score(xtest, ytest)
print("Test score: ", test_score)

Test score:  0.9272


/usr/local/lib/python3.9/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but SGDClassifier was fitted with feature names
  warnings.warn(


Let's evaluate the classifier by examining the  `confusion_matrix`.

In [18]:
ypred = clf2.predict(xtest)
cm = confusion_matrix(ytest, ypred)
print(cm) 

[[2285   24  152]
 [  46 2350   71]
 [  87  166 2319]]


/usr/local/lib/python3.9/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but SGDClassifier was fitted with feature names
  warnings.warn(


In [19]:
cr = classification_report(ytest, ypred)
print(cr) 

              precision    recall  f1-score   support

           0       0.94      0.93      0.94      2461
           1       0.93      0.95      0.94      2467
           2       0.91      0.90      0.91      2572

    accuracy                           0.93      7500
   macro avg       0.93      0.93      0.93      7500
weighted avg       0.93      0.93      0.93      7500



Apart from `SGDClassifier`, we can also train `Perceptron()`, `MultinomialNB()` and `BernoulliNB()` in a similar manner.



---



# **Incremental Preprocessing Example**

## `CountVectorizer` vs `HashingVectorizer`

Vectorizers are used to convert a collection of text documents to a vector representation, thus helping in preprocessing them before applying any model on these text documents. 

`CountVectorizer` and `HashingVectorizer` both perform the task of vectorizing the text documents. However, there are some differences among them. 


One difference is that `HashingVectorizer` does not store the resulting vocabulary (i.e. the unique tokens). Hence, it can be used to learn from data that does not fit into the computer’s main memory. Each mini-batch is vectorized using `HashingVectorizer` so as to guarantee that the input space of the estimator has always the same dimensionality.

With `HashingVectorizer`, each token directly maps to a pre-defined column position in a matrix. For example, if there are 100 columns in the resultant (vectorized) matrix, each token (word) maps to 1 of the 100 columns. The mapping between the word and the position in matrix is done using hashing. 

In other words, in `HashingVectorizer`, each token transforms to a column position instead of adding to the vocabulary. Not storing the vocabulary is useful while handling large data sets. This is because holding a huge token vocabulary comprising of millions of words may be a challenege when the memory is limited.

Since `HashingVectorizer` does not store vocabulary, its object not only takes lesser space, it also alleviates any dependence with function calls performed on the previous chunk of data in case of incremental learning.

###Example

Let us take some sample text documents and vectorize them, first using CountVectorizer and then HashingVectorizer.

In [20]:
text_documents = ['The well-known saying an apple a day keeps the doctor away has a very straightforward, literal meaning, that the eating of fruit maintains good health.',
                  'The proverb first appeared in print in 1866 and over 150 years later is advice that we still pass down through generations.', 
                  'British apples are one of the nations best loved fruit and according to Great British Apples, we consume around 122,000 tonnes of them each year.', 
                  'But what are the health benefits, and do they really keep the doctor away?']


### 1. CountVectorizer

We will first import the library and then create an object of CountVectorizer class. 

In [21]:
from sklearn.feature_extraction.text import CountVectorizer
c_vectorizer = CountVectorizer()

We will now use this object to vectorize the input text documents using the function `fit_transform()`.

In [22]:
X_c = c_vectorizer.fit_transform(text_documents)

In [23]:
X_c.shape

(4, 66)

Here, 66 is the size of the vocabulary.

We can also see the volcabulary using `vocabulary_` attribute.


In [24]:
c_vectorizer.vocabulary_

{'the': 54,
 'well': 62,
 'known': 36,
 'saying': 50,
 'an': 6,
 'apple': 9,
 'day': 19,
 'keeps': 35,
 'doctor': 21,
 'away': 13,
 'has': 30,
 'very': 60,
 'straightforward': 52,
 'literal': 38,
 'meaning': 41,
 'that': 53,
 'eating': 24,
 'of': 43,
 'fruit': 26,
 'maintains': 40,
 'good': 28,
 'health': 31,
 'proverb': 48,
 'first': 25,
 'appeared': 8,
 'in': 32,
 'print': 47,
 '1866': 3,
 'and': 7,
 'over': 45,
 '150': 2,
 'years': 65,
 'later': 37,
 'is': 33,
 'advice': 5,
 'we': 61,
 'still': 51,
 'pass': 46,
 'down': 22,
 'through': 57,
 'generations': 27,
 'british': 16,
 'apples': 10,
 'are': 11,
 'one': 44,
 'nations': 42,
 'best': 15,
 'loved': 39,
 'according': 4,
 'to': 58,
 'great': 29,
 'consume': 18,
 'around': 12,
 '122': 1,
 '000': 0,
 'tonnes': 59,
 'them': 55,
 'each': 23,
 'year': 64,
 'but': 17,
 'what': 63,
 'benefits': 14,
 'do': 20,
 'they': 56,
 'really': 49,
 'keep': 34}

Following is the representation of four text documents.

In [26]:
print(X_c)

  (0, 54)	3
  (0, 62)	1
  (0, 36)	1
  (0, 50)	1
  (0, 6)	1
  (0, 9)	1
  (0, 19)	1
  (0, 35)	1
  (0, 21)	1
  (0, 13)	1
  (0, 30)	1
  (0, 60)	1
  (0, 52)	1
  (0, 38)	1
  (0, 41)	1
  (0, 53)	1
  (0, 24)	1
  (0, 43)	1
  (0, 26)	1
  (0, 40)	1
  (0, 28)	1
  (0, 31)	1
  (1, 54)	1
  (1, 53)	1
  (1, 48)	1
  :	:
  (2, 39)	1
  (2, 4)	1
  (2, 58)	1
  (2, 29)	1
  (2, 18)	1
  (2, 12)	1
  (2, 1)	1
  (2, 0)	1
  (2, 59)	1
  (2, 55)	1
  (2, 23)	1
  (2, 64)	1
  (3, 54)	2
  (3, 21)	1
  (3, 13)	1
  (3, 31)	1
  (3, 7)	1
  (3, 11)	1
  (3, 17)	1
  (3, 63)	1
  (3, 14)	1
  (3, 20)	1
  (3, 56)	1
  (3, 49)	1
  (3, 34)	1




---

### HashingVectorizer
Let us now see how `HashingVectorizer` is different from `CountVectorizer`.

We will create an object of HashingVectorizer. While creating the object, we need to specify the number of features we wish to have in the feature matrix. 

In [27]:
from sklearn.feature_extraction.text import HashingVectorizer

Let us create an object of `HashingVectorizer` class. An important parameter of this class is `n_features`. It declares the number of features (columns) in the output feature matrix. 

Note: Small numbers of features are likely to cause hash collisions, but large numbers will cause larger coefficient dimensions in linear learners.

In [28]:
h_vectorizer= HashingVectorizer(n_features=50) 

Let's perform hashing vectorization with `fit_transform`.

In [29]:
X_h = h_vectorizer.fit_transform(text_documents)


Let us examine the shape of the transformed feature matrix. The number of columns in this matrix is equal to the `n_features` attribute we specified.

In [30]:
X_h.shape

(4, 50)

Let's print the representation of the first example.

In [31]:
print(X_h[0])

  (0, 5)	0.0
  (0, 8)	-0.47140452079103173
  (0, 10)	-0.23570226039551587
  (0, 11)	-0.23570226039551587
  (0, 13)	0.0
  (0, 18)	-0.23570226039551587
  (0, 20)	0.23570226039551587
  (0, 26)	0.0
  (0, 29)	0.23570226039551587
  (0, 33)	0.23570226039551587
  (0, 36)	-0.23570226039551587
  (0, 38)	0.47140452079103173
  (0, 39)	-0.23570226039551587
  (0, 45)	-0.23570226039551587
  (0, 46)	0.23570226039551587


Overall, `HashingVectorizer` is a good choice if we are falling short of memory and resources, or we need to perform incremental learning. However, `CountVectorizer` is a good choice if we need to access the actual tokens.



---



#**Combining preprocessing and fitting in Incremental Learning**

###(`HashingVectorizer` along with `SGDClassifier`)



We will now use a dataset containing a textual feature that requires preprocessing using a vectorizer. Since we wish to perform incremental learning using `partial_fit()`, we will preprocess (i.e., vectorize) the dataset feature using `HashingVectorizer` and then we will incrementally fit it. 

### 1. Downloading the dataset

Below, we download a dataset from UCI ML datasets' library. (Instead of downloading, unzipping and then reading, we are directly reading the zipped csv file. For that purpose, we are making use of `urllib.request`, `BytesIO` and `TextIOWrapper` classes.)

This is a sentiment analysis dataset. There are only two columns in the dataset. One for the textual review and the other for the sentiment.

In [32]:
import pandas as pd
from io import StringIO, BytesIO, TextIOWrapper
from zipfile import ZipFile
import urllib.request

resp = urllib.request.urlopen('https://archive.ics.uci.edu/ml/machine-learning-databases/00331/sentiment%20labelled%20sentences.zip')
zipfile = ZipFile(BytesIO(resp.read()))

data = TextIOWrapper(zipfile.open('sentiment labelled sentences/amazon_cells_labelled.txt'), encoding='utf-8')

df = pd.read_csv(data, sep = '\t')
df.columns = ['review', 'sentiment']

##2. Exploring the data set.

Let's explore the dataset a bit.

In [33]:
df.head()

review  sentiment
0                        Good case, Excellent value.          1
1                             Great for the jawbone.          1
2  Tied to charger for conversations lasting more...          0
3                                  The mic is great.          1
4  I have to jiggle the plug to get it to line up...          0

In [34]:
df.tail()

review  sentiment
994  The screen does get smudged easily because it ...          0
995  What a piece of junk.. I lose more calls on th...          0
996                       Item Does Not Match Picture.          0
997  The only thing that disappoint me is the infra...          0
998  You can not answer calls with the unit, never ...          0

In [35]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 999 entries, 0 to 998
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review     999 non-null    object
 1   sentiment  999 non-null    int64 
dtypes: int64(1), object(1)
memory usage: 15.7+ KB


In [36]:
df.describe()

sentiment
count  999.000000
mean     0.500501
std      0.500250
min      0.000000
25%      0.000000
50%      1.000000
75%      1.000000
max      1.000000

In [37]:
df.loc[:, 'sentiment'].unique()

array([1, 0])

As we can see, 
- There are 999 samples in the dataset. 
- The possible classes for sentiment are 1 and 0.

##4. Splitting data into train and test

In [38]:
from sklearn.model_selection import train_test_split

In [39]:
X = df.loc[:, 'review']

In [40]:
y= df.loc[:, 'sentiment']

In [41]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2)

In [42]:
X_train.shape

(799,)

In [43]:
y_train.shape

(799,)

## 5. Preprocessing

Since the data is textual, we need to vectorize it. In order to perform incremental learning, we will use HashingVectorizer.

In [46]:
from sklearn.feature_extraction.text import HashingVectorizer
vectorizer = HashingVectorizer()

##6. Creating an instance of the SGDClassifier

In [47]:
from sklearn.linear_model import SGDClassifier
classifier = SGDClassifier(penalty='l2',loss='hinge')

## 7. Iteration 1 of partial_fit()

We will assume we do not have sufficient memory to handle all the 799 samples in one go for training purpose. So, we will take the first 400 samples from teh training data and `partial_fit` our classifier.

Another use case of partial_fit here could also be a scenario where we only have 400 samples available at a time. So, we fit our classifier with them. However, we `partial_fit` it, to have the possibility of training it wirth more data later whenever that becomes available.


In [48]:
X_train_part1_hashed = vectorizer.fit_transform(X_train[0:400])
y_train_part1 = y_train[0:400]


In [50]:
all_classes = np.unique(df.loc[:, 'sentiment']) #we need to mention all classes in the first iteration of partial_fit()

In [51]:
classifier.partial_fit(X_train_part1_hashed, y_train_part1, classes=all_classes)

SGDClassifier()

Let us now use this classifier on our test data that we had kept aside earlier.

In [52]:
X_test_hashed = vectorizer.transform(X_test) #first we will have to preprocess the X_test with the same vectorizer that was fit on train data.

In [53]:
test_score = classifier.score(X_test_hashed, y_test)
print("Test score: ", test_score)

Test score:  0.74


Note: We can also store this classifier using pickle object and can access it later.

# 8. Iteration 2 of partial_fit()

We will now assume that more data became available. So, we will fit the same classifier with more data and observe if our test score improves.

In [54]:
X_train_part2_hashed = vectorizer.fit_transform(X_train[400:])
y_train_part2 = y_train[400:]


In [55]:
classifier.partial_fit(X_train_part2_hashed, y_train_part2)

SGDClassifier()

In [56]:
test_score = classifier.score(X_test_hashed, y_test)
print("Test score: ", test_score)

Test score:  0.75


We see that our test score has improved after we fed more data to the classifier in the second iteration of `partial_fit()`.

For a more elaborate example, refer: https://scikit-learn.org/stable/auto_examples/applications/plot_out_of_core_classification.html#sphx-glr-auto-examples-applications-plot-out-of-core-classification-py